### PR TITLE
Add type hints to _nacl.py

### LINKED ISSUE #3570: Add type hints to _nacl.py
Add type hints to src/toil/lib/encryption/_nacl.py so it can be checked under mypy during linting.

Refers to #3568.

┆Issue is synchronized with this [Jira Task](https://ucsc-cgl.atlassian.net/browse/TOIL-880)
┆Issue Number: TOIL-880

### PR DESCRIPTION
**Pull request by jonathanxu18**
Adds type hints to src/toil/lib/encryption/_nacl.py so it can be checked under mypy during linting.

Should fix #3570.

Refers to #3568.

## Changelog Entry
To be copied to the [draft changelog](https://github.com/DataBiosphere/toil/wiki/Draft-Changelog) by merger:

Type hints added to: 
* `encrypt()`
* `decrypt()`

## Reviewer Checklist

 * [ ] Make sure it is coming from `issues/XXXX-fix-the-thing` in the Toil repo, or from an external repo.
    * [ ] If it is coming from an external repo, make sure to pull it in for CI with:
        ```
        contrib/admin/test-pr otheruser theirbranchname issues/XXXX-fix-the-thing
        ```
    * [ ] If there is no associated issue, [create one](https://github.com/DataBiosphere/toil/issues/new).
* [ ] Read through the code changes. Make sure that it doesn't have:
    * [ ] Addition of trailing whitespace.
    * [ ] New variable or member names in `camelCase` that want to be in `snake_case`.
    * [ ] New functions without [type hints](https://docs.python.org/3/library/typing.html).
    * [ ] New functions or classes without informative docstrings.
    * [ ] Changes to semantics not reflected in the relevant docstrings.
    * [ ] New or changed command line options for Toil workflows that are not reflected in `docs/running/cliOptions.rst`
    * [ ] New features without tests.
* [ ] Comment on the lines of code where problems exist with a review comment. You can shift-click the line numbers in the diff to select multiple lines.
* [ ] Finish the review with an overall description of your opinion.

## Merger Checklist

* [ ] Make sure the PR passes tests.
* [ ] Make sure the PR has been reviewed **since its last modification**. If not, review it.
* [ ] Merge with the Github "Squash and merge" feature.
    * [ ] If there are multiple authors' commits, add [Co-authored-by](https://github.blog/2018-01-29-commit-together-with-co-authors/) to give credit to all contributing authors.
* [ ] Copy its recommended changelog entry to the [Draft Changelog](https://github.com/DataBiosphere/toil/wiki/Draft-Changelog).
* [ ] Append the issue number in parentheses to the changelog entry.

### CODE REVIEW

**Comment by DailyDreaming** [comment]
@jonathanxu18 The mypy tests are failing:

```
$ make mypy
/builds/databiosphere/toil/contrib/admin/mypy-with-ignore.py
src/toil/lib/encryption/_nacl.py:15: error: Skipping analyzing 'nacl': found module but no type hints or library stubs
src/toil/lib/encryption/_nacl.py:15: note: See https://mypy.readthedocs.io/en/latest/running_mypy.html#missing-imports
src/toil/lib/encryption/_nacl.py:16: error: Skipping analyzing 'nacl.secret': found module but no type hints or library stubs
src/toil/lib/encryption/_nacl.py:96: error: Returning Any from function declared to return "bytes"
Found 3 errors in 1 file (checked 3 source files)
```

I'll leave it to you to decide how to fix these.  Mypy may also be too strict on some topics as well, so if you think something might be going too far (like adding an entire file to satisfy mypy or something), please comment or bring it up at the next stand up.

**Review by DailyDreaming — COMMENTED** [review]
See above.  ^

**Comment by mr-c** [comment]
> I'll leave it to you to decide how to fix these.

https://github.com/python/typeshed/search?q=pynacl comes up empty, so I recommend

```
mkdir -p contib/typeshed
stubgen -p nacl -o contrib/typeshed
```

And the adding `MYPYPATH=contrib/typeshed` to the `mypy` invocation.

**Review by DailyDreaming — COMMENTED** [review]
Adding 32 files seems kind of intense.  @jonathanxu18 Is there any way around this?

File: contrib/typeshed/nacl/secret.pyi [hunk 1]
```diff
@@ -0,0 +1,13 @@
+from nacl import encoding as encoding
+from nacl.utils import EncryptedMessage as EncryptedMessage, StringFixer as StringFixer, random as random
+from typing import Any, Optional
```
Comment (mr-c) [inline contrib/typeshed/nacl/secret.pyi:3]:
```suggestion
from typing import Any, Optional, SupportsBytes
```

File: contrib/typeshed/nacl/secret.pyi [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+class SecretBox(encoding.Encodable, StringFixer):
+    KEY_SIZE: Any = ...
+    NONCE_SIZE: Any = ...
```
Comment (mr-c) [inline contrib/typeshed/nacl/secret.pyi:7]:
```suggestion
    KEY_SIZE: int = ...
    NONCE_SIZE: int = ...
```

File: contrib/typeshed/nacl/utils.pyi [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+class EncryptedMessage(bytes):
+    @property
+    def nonce(self): ...
+    @property
+    def ciphertext(self): ...
+
+class StringFixer: ...
+
+def bytes_as_string(bytes_in: Any): ...
+def random(size: int = ...): ...
```
Comment (mr-c) [inline contrib/typeshed/nacl/utils.pyi:12]:
```suggestion
def random(size: int = ...) -> str: ...
```

File: contrib/typeshed/nacl/secret.pyi [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+    MACBYTES: Any = ...
+    MESSAGEBYTES_MAX: Any = ...
+    def __init__(self, key: Any, encoder: Any = ...) -> None: ...
+    def __bytes__(self): ...
+    def encrypt(self, plaintext: bytes, nonce: Optional[Any] = ..., encoder: Any = ...) -> bytes: ...
+    def decrypt(self, ciphertext: bytes, nonce: Optional[Any] = ..., encoder: Any = ...) -> bytes: ...
```
Comment (mr-c) [inline contrib/typeshed/nacl/secret.pyi:13]:
```suggestion
    def encrypt(self, plaintext: bytes, nonce: Optional[bytes] = ..., encoder: Any = ...) -> bytes: ...
    def decrypt(self, ciphertext: bytes, nonce: Optional[bytes] = ..., encoder: Any = ...) -> bytes: ...
```

File: contrib/typeshed/nacl/secret.pyi [hunk 5]
```diff
@@ -0,0 +1,13 @@
+from nacl import encoding as encoding
+from nacl.utils import EncryptedMessage as EncryptedMessage, StringFixer as StringFixer, random as random
+from typing import Any, Optional
+
+class SecretBox(encoding.Encodable, StringFixer):
+    KEY_SIZE: Any = ...
+    NONCE_SIZE: Any = ...
+    MACBYTES: Any = ...
+    MESSAGEBYTES_MAX: Any = ...
+    def __init__(self, key: Any, encoder: Any = ...) -> None: ...
```
Comment (mr-c) [inline contrib/typeshed/nacl/secret.pyi:10]:
```suggestion
    def __init__(self, key: SupportsBytes, encoder: Any = ...) -> None: ...
```

File: contrib/typeshed/nacl/utils.pyi [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+
+class EncryptedMessage(bytes):
+    @property
+    def nonce(self): ...
+    @property
+    def ciphertext(self): ...
+
+class StringFixer: ...
+
+def bytes_as_string(bytes_in: Any): ...
```
Comment (mr-c) [inline contrib/typeshed/nacl/utils.pyi:11]: Not used by Toil

```suggestion
```

File: contrib/typeshed/nacl/utils.pyi [hunk 7]
```diff
@@ -0,0 +1,12 @@
+from typing import Any
+
+class EncryptedMessage(bytes):
+    @property
+    def nonce(self): ...
+    @property
+    def ciphertext(self): ...
```
Comment (mr-c) [inline contrib/typeshed/nacl/utils.pyi:7]: Not used by Toil
```suggestion
```

File: contrib/typeshed/nacl/secret.pyi [hunk 8]
```diff
@@ -0,0 +1,13 @@
+from nacl import encoding as encoding
+from nacl.utils import EncryptedMessage as EncryptedMessage, StringFixer as StringFixer, random as random
```
Comment (mr-c) [inline contrib/typeshed/nacl/secret.pyi:2]:
```suggestion
from nacl.utils import StringFixer, random
```
EncryptedMessage is not used by Toil

**Review by mr-c — CHANGES_REQUESTED** [review]
Some suggestions to simplify & improve the type hints

File: contrib/typeshed/nacl/utils.pyi (hunk 3 above)
Comment (jonathanxu18) [reply contrib/typeshed/nacl/utils.pyi:12]: Mypy was complaining when encrypt() is called after this change but returning bytes instead seems to be working.

File: contrib/typeshed/nacl/secret.pyi (hunk 5 above)
Comment (jonathanxu18) [reply contrib/typeshed/nacl/secret.pyi]: SupportBytes also was producing an error for mypy but switching to bytes seems to be working.

**Review by mr-c — APPROVED** [review]

**Pull request merged**